### DISCUSSION TITLE
Determine how long UploadFile took to upload.

### DISCUSSION THREAD

**Discussion by MarioIshac** [fastapi/fastapi#7000] [Questions] [answered]
Given this example: 
```
from fastapi import FastAPI, UploadFile, File

app = FastAPI()

@app.post("/a")
async def my_function(file: UploadFile = File(...)):
    _ = await file.read()
```

`file.read()` either reads the file from disk or memory (depending on whether the underling spooled file is in memory to begin with or not), but `my_function` is only invoked once `UploadFile` finishes uploading to the server. Is there a way to know how long it took the client to upload `UploadFile`?

**Comment by valentin994** [comment]
Do you get your response before the upload finishes? If not this should work :D.

```
@app.middleware("http")
async def add_process_time_header(request: Request, call_next):
    start_time = time.time()
    response = await call_next(request)
    process_time = time.time() - start_time
    response.headers["X-Process-Time"] = str(process_time)
    return response
```

**Comment by MarioIshac** [comment]
> Do you get your response before the upload finishes? If not this should work :D.
> 
> ```
> @app.middleware("http")
> async def add_process_time_header(request: Request, call_next):
>     start_time = time.time()
>     response = await call_next(request)
>     process_time = time.time() - start_time
>     response.headers["X-Process-Time"] = str(process_time)
>     return response
> ```

Ah I should have given a better example, but thanks for your response. I already have something like this, but this adds the processing time for the entire request. In reality, my request contains the `UploadFile` plus lots of calls within the request itself. I'd like to know the time that the `UploadFile` takes to upload, isolated from the time the rest of the request takes.

**Comment by valentin994** [comment]
> > Do you get your response before the upload finishes? If not this should work :D.
> > 
> > ```
> > @app.middleware("http")
> > async def add_process_time_header(request: Request, call_next):
> >     start_time = time.time()
> >     response = await call_next(request)
> >     process_time = time.time() - start_time
> >     response.headers["X-Process-Time"] = str(process_time)
> >     return response
> > ```
> 
> Ah I should have given a better example, but thanks for your response. I already have something like this, but this adds the processing time for the entire request. In reality, my request contains the `UploadFile` plus lots of calls within the request itself. I'd like to know the time that the `UploadFile` takes to upload, isolated from the time the rest of the request takes.

I do think the start time would be right, you would just need to stamp the end time earlier. So you could subtract the time that the method from the first response gave and use the start of my_function time to get your file upload time.

**Comment by MarioIshac** [comment]
@valentin994 Alright thanks for the followup, I've been looking into how I can "return" the time recorded at start of `my_function` out of `my_function`. Would I sent it through `request.state` (after declaring a `request: Request` as a parameter)? Or is there a more intuitive way?

**Comment by krishnardt** [comment]
UploadFile purpose is just for the validation if the object being sending is a file or not.

[https://github.com/tiangolo/fastapi/blob/master/fastapi/datastructures.py](url)
you can check this link for further details.
to understand how much time it is taking to check... you can tweak it. like below..

```
class UploadFile(StarletteUploadFile):
    @classmethod
    def __get_validators__(cls: Type["UploadFile"]) -> Iterable[Callable[..., Any]]:
        start_time = time.time()
        print(start_time)
        yield cls.validate
        end_time = time.time()
        print(end_time)
        print(end_time-start_time)

    @classmethod
    def validate(cls: Type["UploadFile"], v: Any) -> Any:
        if not isinstance(v, StarletteUploadFile):
            raise ValueError(f"Expected UploadFile, received: {type(v)}")
        return v

```

This will tells us how much time taken to vaildate all files.

**Comment by krishnardt** [comment]
> @valentin994 Alright thanks for the followup, I've been looking into how I can "return" the time recorded at start of `my_function` out of `my_function`. Would I sent it through `request.state` (after declaring a `request: Request` as a parameter)? Or is there a more intuitive way?

you may refer this issue for the solution.
https://github.com/tiangolo/fastapi/issues/2727
@ycd @Kludex solutions may give you a hint.

**Comment by MarioIshac** [comment, marked as answer]
Thanks guys for all the help, I've come up with this solution:

```
@app.post("/")
async def post_file(request: Request, file: UploadFile = File(...)):
    request.state.upload_end_time = time.time()

    # Proceed as normal

@app.middleware("http")
async def get_timed_response(request: Request, call_next: Callable[[Request], Awaitable[Response]]) -> Response:
    start_time = time.time()
    response = await call_next(request)

    if upload_end_time := getattr(request.state, "upload_end_time", None):
        upload_time = upload_end_time - start_time
        response.headers["X-Upload-Time"] = str(upload_time)

    return response
```

I have to avoid using `Depends(...)` in the parameter list and instead do it within the request itself, but this is small downside considering that I only have one request taking an `UploadFile` for which I care about the upload time for.

**Comment by MarioIshac** [comment]
For anyone that comes across this issue, note that the above solution is accurate only if FastAPI is not behind a reverse proxy or the reverse proxy does no file buffering. I ended up scrapping trying to do this from application server side and instead monitor upload times under my reverse proxy. For Nginx, client can do `$request_time - $upload_response_time` (I send them as two separate headers) to get their accurate upload time.

**Comment by tiangolo** [comment]
Thanks for the help here @valentin994 and @krishnardt ! 👏  🙇 

Thanks for reporting back and closing the issue @MarioIshac  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
